### ISSUE TITLE
linking problem, unknown symbol `strdup'

### ISSUE THREAD

**Issue by LoschmidtParadox** [bsl/bindings-GLFW#27]
Hi, 
I'm having an issue with strdup which I noticed was in a prior issue (https://github.com/bsl/bindings-GLFW/issues/11)

```
[1 of 3] Compiling Graphics.UI.GLFW.Types ( Graphics\UI\GLFW\Types.hs, dist\build\Graphics\UI\GLFW\Types.o )
Loading package ghc-prim ... linking ... done.
Loading package integer-gmp ... linking ... done.
Loading package base ... linking ... done.
Loading package bindings-DSL-1.0.22 ... linking ... done.
Loading package bindings-GLFW-3.1.1.2 ... linking ... ghc.exe: unable to load package `bindings-GLFW-3.1.1.2'
ghc.exe: warning: _vsnprintf from msvcrt is linked instead of __imp__vsnprintf
ghc.exe: C:\Users\Pawel\AppData\Roaming\cabal\x86_64-windows-ghc-7.8.3\bindings-GLFW-3.1.1.2\HSbindings-GLFW-3.1.1.2.o: unknown symbol `strdup'
Failed to install GLFW-b-1.4.7.2
cabal.exe: Error: some packages failed to install:
GLFW-b-1.4.7.2 failed during the building phase. The exception was:
ExitFailure 1

```
Here are my versions:

```
$ ghc --version
The Glorious Glasgow Haskell Compilation System, version 7.8.3

$ ghc --print-target-platform
x86_64-unknown-mingw32

$ ghc --print-build-platform
x86_64-unknown-mingw32

$ cabal --version
cabal-install version 1.22.0.1
using version 1.22.1.1 of the Cabal library

```

I checked if any earlier versions work, and apparently bindings-GLFW-3.0.3.3 works with GLFW-b-1.4.6 (or at the very least it compiles and installs while throwing a bunch of warnings).

Could it be that when the workaround was moved to win32_platform.h that it stopped working for 64-bit windows?

**Comment by LoschmidtParadox** [comment]
I've tried it out and I can confirm that the workaround compiles and successfully runs gears demo code.
I now just get a ton of warnings, as I did with compiling earlier versions.

for example:
```
Loading package bindings-GLFW-3.1.1.2 ... linking ... <interactive>: warning: _vsnprintf from msvcrt is linked instead of __imp__vsnprintf
<interactive>: warning: _strdup from msvcrt is linked instead of __imp__strdup
<interactive>: warning: MultiByteToWideChar from kernel32 is linked instead of __imp_MultiByteToWideChar
<interactive>: warning: WideCharToMultiByte from kernel32 is linked instead of __imp_WideCharToMultiByte
<interactive>: warning: SystemParametersInfoW from user32 is linked instead of __imp_SystemParametersInfoW
<interactive>: warning: LoadLibraryW from kernel32 is linked instead of __imp_LoadLibraryW
```

**Comment by gglouser** [comment]
I found another project that had this issue: https://github.com/snoyberg/yaml/issues/44

There is a ticket for it on GHC's trac (https://ghc.haskell.org/trac/ghc/ticket/9215), which leads to another ticket (https://ghc.haskell.org/trac/ghc/ticket/9218) about updating the version of MinGW shipped with GHC. There appears to be a patch, but it did not make it into 7.10.1. Currently, it is slated for 7.12.1. So, it might be necessary to work around this until that GHC fix.

**Comment by schell** [comment]
@gglouser Great sleuthing! If that's the case then I feel better about adding the fix to the glfw source.

**Comment by schell** [comment]
I've pushed the fix to hackage.

**Comment by SwiftsNamesake** [comment]
I'm having the same issue with GHC 8.0.1.

**Comment by cloudhead** [comment]
Supposedly this should be fixed in 8.0.2: https://github.com/bsl/bindings-GLFW/issues/48#issuecomment-263063191
